---
layout: single
title: "Using Autoencoders for Breast Cancer Classification"
excerpt: "Data analysis in Python by recreating published figures using available data that provided a unique learning opportunity to build a comprehensive skill set for tackling contemporary challenges at the intersection of biology and artificial intelligence."

classes: wide
date: 2023-12-01
last_modified_at: 2024-02-18
tags:
  - machine learning
  - python
  - tensorflow
category:
  - academic
---
<i class="fab fa-github"></i>[Project Repository](https://github.com/Jstream11/Breast_Cancer_Autoencoder_Project.git)

*The data for this project was taken from Wang S, Lee D (2023) Identifying prognostic subgroups of luminal-A breast cancer using deep autoencoders and gene expressions. PLOS Computational Biology 19(5): e1011197. [https://doi.org/10.1371/journal.pcbi.1011197](https://doi.org/10.1371/journal.pcbi.1011197)*

<img src="/assets/images/AutoencoderProj/Autoencoder_Clustering.png" alt="t-SNE plot">

**Fig 1. The t-SNE (t-distributed Stochastic Neighbor Embedding) plot of 64-dimensional latent features.** The t-SNE plot (dimension size = 2) of latent features generated from the deep autoencoders of 679 METABRIC luminal-A breast cancer samples. The samples assigned to the BPS-LumA (the better prognostic subgroup) and WPS-LumA (the worse prognostic subgroup) colored as brown and blue, respectively.


<img src="/assets/images/AutoencoderProj/Autoencoder_SurvivalCurve.png" alt="Kaplan-Meier Curve">

**Fig 2. The Kaplan-Meier survival curve of 679 METABRIC luminal-A breast cancer samples.** (A) The survival curves were calculated using the clustering data resulting from the latent features generated by the autoencoder. The brown and blue curves indicate the BPS-LumA and the WPS-LumA, respectively. The x-axis refers to recurrence-free survival months and the y-axis refers survival probability. (B) The survival curves were calculated using the published subgroup information from the supplemental data (Table S4).

### The Goal

---
Practice data analysis in Python by recreating published figures using available data. The selection of this paper for replication was driven by the desire to not only work with expression data but also the intricacies presented by machine learning methodologies. The inclusion of advanced machine learning techniques, such as the implementation of deep autoencoders in subgroup identification, added a layer of complexity and nuance to the replication process. This multidimensional approach provided a unique opportunity to build a comprehensive skill set for tackling contemporary challenges at the intersection of biology and artificial intelligence.

### The Idea

---

The focal point of this research is on breast cancer as a heterogeneous disease, with distinct subtypes necessitating tailored high throughput therapeutic strategies due to varying subtypes resulting from different molecular mechanisms. As the most common subtype, luminal-A breast cancer, presents a unique challenge: despite its high expression of hormone receptors, resistance to endocrine therapies complicates treatment outcomes. The motivation behind this study is rooted in the need for a precise stratification method that can accurately predict prognosis and facilitate personalized medicine approaches.

Utilizing deep autoencoders, the study delves into gene expressions from a cohort of 679 luminal-A breast cancer samples within the METABRIC dataset. The result of this approach is the identification of two distinct prognostic subgroups, denoted as BPS-LumA (the better prognostic subgroup) and WPS-LumA (the worse prognostic subgroup).

K-Means clustering, based on latent features extracted by the deep autoencoders, and subsequent Kaplan-Meier survival analysis, unravel a significant difference in prognoses between BPS-LumA and WPS-LumA (p-value = 5.82E-05). The validation process, conducted with an independent dataset comprising 415 samples from the TCGA BRCA dataset, further substantiates the identified prognostic differences (p-value = 0.004).

Notably, the study highlights the superiority of latent features over conventional methods, such as gene expression profiles and Principal Component Analysis (PCA). This underscores the efficacy of the deep autoencoder approach in capturing essential information that contributes to a more nuanced understanding of the disease's prognosis.

### The Design

---

*Data Preprocessing and Renormalization*

The METABRIC and TCGA data was obtained from cBioPortal. The expression data used from the METABRIC dataset was normalized microarray zscores while the TCGA expression data was normalized RNAseq zscores. The samples were filtered based on the supplemental information in the paper (Tables S4 and S5). The genes were filtered based on the top 5000 expressed gene information from the supplemental information (Table S1).

The supplemental data did not contain the data from the autoencoders or the renormalized expression data. The protocol from their published GitHub repository was used to renormalize the METABRIC and TCGA expression data sets. In order to get the renormalization methods to work, additional preprocessing steps were taken. First, the gene features of both datasets were compared to ensure both the METABRIC and TCGA expression data represented the same genes. Next, any duplicate genes in the data sets were removed. Finally, both data sets were renormalized using MinMaxScaler().

Next, training and testing features were created to input into the autoencoder. The training features were created as a numpy array with the renormalized METABRIC dataset, filtering based on the top 5000 gene list. The testing set was made similarly but with the TCGA dataset. It is important to note that, to have datasets of the same feature size, after taking the overlap of genes between the breast cancer datasets and then filtering for the top genes, the resulting set was 4986 genes, not 5000 genes. This might reflect a difference in the clustering and survival analysis further downstream.

Before autoencoder development, there was need to interpolate the training features to fill NA values in the training set. This was not done in the published code and could also introduce variance.

*Autoencoder*

The autoencoder was developed exactly as described in the published repository using TensorFlow and Keras. The autoencoder consists of five layers: an input layer, three hidden layers of the same size (set to a dimension of 64), and an output layer. The Rectified Linear Unit (ReLU) activation function is used for the hidden layers, except for the output layer, which uses the sigmoid activation function. The code creates three models: the full autoencoder model, an encoder model that maps input features to their encoded representations, and a decoder model for reconstructing the input features from the encoded representations. The autoencoder is compiled using the Adam optimizer and mean squared error (MSE) as the loss function. The training process involves fitting the autoencoder to the training features, using the same features as both input and target output, over 100 epochs with a batch size of 16, with the testing features used as validation data. To generate the encoded representations (latent features) of the training features, the predict() method is applied to the encoder model, transforming the METABRIC data into a more condensed and abstract representation. The resulting metabric_latent variable holds these encoded features, capturing essential patterns or characteristics of the original data in a more compact form, suitable for various downstream tasks in machine learning.

*Clustering and tSNE*

Using the KMeans algorithm, specifying 2 clusters, the latent features from METABRIC are clustered. Subsequently, t-Distributed Stochastic Neighbor Embedding (t-SNE) is employed to project the 64-dimensional latent features into a two-dimensional space for visualization. The resulting t-SNE plot is displayed, where data points are color-coded based on the clusters assigned by KMeans.

*Kaplan-Meier Survival Curve*

Survival analysis on the METABRIC dataset was performed based on the data, assessing their recurrence-free survival (RFS) using Kaplan-Meier survival curves. RFS and survival times are extracted from metadata, obtained from cBioPortal. The KaplanMeierFitter from the lifelines library is then used to estimate survival functions for each cluster, and the curves are plotted on a single graph. The multivariate_logrank_test assessed whether survival curves differ significantly among groups, providing a p-value. Additionally, pairwise_logrank_test was employed for pairwise comparisons between groups. The resulting Kaplan-Meier curves were visualized, showing distinct survival patterns for clusters labeled as "BPS-LumA" and "WPS-LumA." The plot included the calculated p-value and provided a comprehensive overview of the survival probabilities over time for the clustered luminal-A breast cancer samples.

### The Results

---

In replicating the findings from the paper, I successfully recreated the tSNE plot with two clusters, aligning with the BPS and WPS subgroups identified in the original research (Fig 1). However, when generating the survival curve for these subgroups, I observed that while the curves followed the same trend as in the published figure, I could not achieve a similar significance reported in the paper (Fig 2a). This lack of replication could be attributed to the inherent challenges in reproducing the established code, requiring additional preprocessing steps on my end. Notably, I encountered difficulty obtaining complete expression data for the top 5000 genes in both the TCGA and METABRIC datasets, and I had to address missing values, aspects not explicitly covered in the original protocols. I hypothesize that subtle differences in the source data, particularly in renormalization procedures, might contribute to variations in the latent features, thus influencing the significance observed in the Kaplan-Meier survival curve results. As an alternative approach, I also plotted the survival curve based on the accepted published subgroups (Table S4), utilizing the samples classified as BPS and WPS by their autoencoder. In this case, I successfully replicated the survival curve plot with the correct p-value (Fig 2a). This underscores the potential influence of nuanced data variations and preprocessing steps in achieving precise replication of complex analyses.
